### ISSUE TITLE
lx_fs: Implementation of unlink

### ISSUE THREAD

**Issue by nfeske** [genodelabs/genode#1120]
Currently, lx_fs lacks support for unlink. Since @ssumpf added the test of unlink to the fs-test, we get the following output:

```
[init -> lx_fs] unlink not implemented
[init -> test-libc_vfs] unlink(file_name) succeeded
[init -> test-libc_vfs] calling stat(file_name, &stat_buf) file_name=test.tst
[init -> test-libc_vfs] stat(file_name, &stat_buf) failed, ret=0, errno=0
[init] virtual void Genode::Child_policy::exit(int): child exited with exit value -1
```

**Comment by ehmry** [comment]
This test passes for me now, though I'm not sure I covered everything.

**Comment by chelmuth** [comment]
This feature is still not merged and I can't remember which reservations lead to this. Was it simply forgotten? @ehmry would you agree a rebase of b48642d866b5a5f9c210fa9fdf7b43b0a129554b should be merged to finally close this issue? Are you willing to address this?

**Comment by nfeske** [comment]
The potential issue I see is the need for sanitizing path arguments.

In particular, what happens when passing something like `../../../.bashrc` to the unlink operation of lx_fs?

Symlinks are also risky: an absolute path specified by an lx_fs client should never end up 1:1 in the symlink inode of the Linux file system. Otherwise the client may become able to read content outside the lx_fs directory. As far as I can see, the write (to symlink) method takes no precaution in this respect.

Granted lx_fs should never be used in scenarios where security matters. But we should still strive to avoid such problems.

The best way to merge this feature with confidence would be a bad-case test that shows that those concerns are covered.

**Comment by ehmry** [comment]
I'm willing to rebase this, but not to maintain it, I have not used lx_fs in a very long time.

**Comment by ehmry** [comment]
Well, unless other people are really using it.

**Comment by chelmuth** [comment]
The reason I commented above that we recently added move() support but I myself also do not use this component. And given the risks mentioned by @nfeske it may be better to leave lx_fs well alone.

**Comment by cnuke** [comment]
W/o having followed the code, what about using `unlinkat(2)` and `symlinkat(2)`?

**Comment by nfeske** [comment]
There are currently no plans to add this feature.

**Comment by sidhussmann** [comment]
@nfeske actually, we are finishing up some missing features in `lx_fs` including unlink. Should we create a new issue or re-open this one?

**Comment by nfeske** [comment]
I'd prefer you re-opening the existing one. So its easier to keep track of the past discussions.

**Comment by sidhussmann** [comment]
@nfeske, that sounds reasonable. However, it seems that Github access control prevents me from doing so.

**Comment by nfeske** [comment]
Here you go. ;-)

**Comment by trimpim** [comment]
We added this in preparation of adding watch to `lx_fs` (#4070), in order to be able to run the `run/lx_fs` test. It adds `unlink` support. Support for `symlinks` is left out because of security considerations.

**Comment by nfeske** [comment]
Fixed in Genode 21.08
